Add site replication docs (#469)
minio/docs: source/replication/site-replication-overview.rst

.. _minio-site-replication-overview:

=========================
Site replication overview
=========================

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 1

> Site replication configures multiple independent MinIO sites or clusters as replicas.

Reviewer: I've been trying to use `MinIO Deployments` as the preferred term over 'clusters' or 'sites', though we should check w/ engineering and make sure we are completely consistent in usage here.

> When configured this way, refer to the replica sites as either "peer sites" or "sites".

Reviewer: I think we can integrate this into the previous sentence and possibly drop one use of 'sites'

> Site replication assumes the use of an external identity provider (IDP).

Reviewer: Q: Does site replication not work with the MinIO internal IDP?

All configured deployments must use the same external IDP.

What replicates across all sites?
---------------------------------

> When enabling site replication on a set of sites, the following changes apply to the other sites:

Reviewer: Try to restructure with either 'replication' or 'synchronization' as the key verb.

Spitballing:

```suggestion
Each MinIO deployment ("replica") in site replication synchronize the following actions:
```

Reviewer: Also - server settings should also synchronize correct? Such as the majority of settings (excluding bucket notification targets) you can set via `mc admin confiig set` ?

Author: No, server settings do not synchronize. It's bucket, bucket settings, objects, object settings, and IAM stuff.

Reviewer: interesting - server settings here also would include the AD/LDAP or OIDC settings you typically set via envvar.

I was hoping for a system where you could configure one deployment and all the others would 'inherit' those configuration settings (minus bucket notification and lifecycle tiering), but this is still something more than what is.

- Creation and deletion of buckets and objects, including

  - Configuration
  - Policy changes and deletions
  - Tags
  - Locks
  - Encryption settings

- Creation and deletion of IAM users, groups, policies, and policy mappings to users or groups (for LDAP users or groups)
- Creation of Security Token Service (STS) credentials for session tokens verifiable from the local ``root`` credentials
- Creation and deletion of service accounts (except those owned by the ``root`` user)
- Changes to bucket features, such as
  
  - :ref:`Policies <minio-policy>`
  - Tags (for example, see :mc-cmd:`mc tag set`)
  - :ref:`Object-Lock <minio-object-locking>` configurations (including retention and legal hold configurations)

>   - :ref:`Encryption configuration <minio-encryption-overview>`

Reviewer: We might merge this up into 25 " Creation, modification, or deletion of..."

Then we can put the references in-line (Don't necessarily need 'for example', we can just link straight).

> Site replication enables bucket versioning for all new and existing buckets on all replicated sites.

Reviewer: Link 'bucket-versioning' to the Versioning overview.

What doesn't replicate across sites?
-------------------------------------

> The following bucket features don't replicate across sites set up for site replication:

Reviewer: Same as above w.r.t. structure

> - Notification configurations

Reviewer: Link these to the relevant reference pages.

"Bucket Notification" configuration (Notification alone might be misread as audit logging or something similar)

- Lifecycle management (ILM) configurations

This allows users who access different sites to define separate notification or ILM settings.

> Order of operations

Reviewer: If this is just for the initial sync we should refocus the title around that.

Author: One site has data, the other(s) do not.
After replication starts, these are the things that do and do not replicate to other sites at the initial and all subsequent syncs.
I added site configuration settings since they do not sync.
And I added links.

-------------------

After enabling site replication, identity and access management (IAM) settings sync in the following order:

#. Policies
#. User accounts (for local users)
   

>    Note: Sites that use OIDC and LDAP don't sync, and they may not have local users.

Reviewer: I wonder if we could tab this out.

For AD/LDAP for example I thought that users/groups were still disabled, as we query them from the AD/LDAP server.

Author: Tabs are a Good Idea (TM) here.

#. Groups 
   
   For MinIO IDP and LDAP managed groups

#. Service accounts

>    Only static accounts with valid claims.

Reviewer: Not sure what 'static accounts' means here - clarify? The 'valid claims' sounds like this is OIDC specific.

Author: Oh, tabs are a Good Idea (TM) here.

   Service accounts for ``root`` don't sync.

#. Policy mapping for synced user accounts

> #. Policy mapping for `Short Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

Reviewer: STS should be Security Token Service.

Site healing
------------

> With site replication enabled, bucket metadata and IAM entries can heal from whichever peer site has the most updated entry.

Reviewer: I assume this is true in general? For any of the items which replication supports (identified above), the MinIO deployment resynchronizes that item from the remote replica with the most updated ("latest") version of that item.

Prerequisites
-------------

> - Only *one* site can have data at the time of setup.

Reviewer: Style note - I try making each pre-req a sub-heading so they are findable and linkable. e.g.

```
Only One Site Can Have Data At Setup
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Details
```

  The other sites must be empty of buckets and objects.

>   After configuring site replication, the data replicates to the other sites.
>   Objects written to any site after setting up replication then replicate to all sites.

Reviewer: I think this could use a bit of rework. Spitballing

```suggestion
  Data from the first site replicates to all others in the configuration as part of the initial setup.
  Once setup completes, :ref:`replicable changes <blah>` made on any site synchronize to all other sites in the configuration.
```

- All sites must have the same deployment credentials (for example, ``MINIO_ROOT_USER``, ``MINIO_ROOT_PASSWORD``)
- All sites must use the same external IDP(s), if any
- For :ref:`SSE-S3 <minio-encryption-sse-s3>` or :ref:`SSE-KMS <minio-encryption-sse-kms>` encryption via KMS, all sites must have access to a central KMS deployment. 
  This can be achieved via a central KES server or multiple KES servers (say one per site) connected via a central KMS (Vault) server.

Configure site replication
--------------------------

.. tab-set::

   .. tab-item:: Console

>       #. Deploy three or more separate MinIO sites, using the same Identity Provider for each site

Reviewer: Maybe link "Deploy" to our deployment tutorial.

Author: Three was a misunderstanding on my part. Two is sufficient, if not wise.

         Only one site can have any buckets or objects on it.
         The other sites must be empty.

>       #. In a browser, access the Console for one of the sites

Reviewer: Q: Does it not have to be the site w/ data,or can you configure from any site?

         For example, ``https://<addressforsite>:9000``
         

>          If one of the sites has content, log in to the Console for that site.

Reviewer: 'Tis a pity we have to be specific, but it's good cya to say "replace ``addressforsite`` with the URL or hostname for your MinIO deployment"

      #. Select **Settings**, then **Site Replication**

         .. image:: /images/minio-console/console-settings-site-replication.png
            :width: 400px
            :alt: MinIO Console menu with the Settings heading expanded to show Site Repilication
            :align: center
      
      #. Select :guilabel:`Add Sites +``

         .. image:: /images/minio-console/console-settings-site-replication-add.png
            :width: 600px
            :alt: MinIO Console's Add Sites for Replication screen
            :align: center

>       #. Make the following entries:

Reviewer: ```suggestion
      #. Complete the requested information for the site:
```

>          :Access Key: `(required)` The user name to use for signing in to each site. Should be the same across all sites.

Reviewer: :thinking: 

So for MinIO IDP this could be any user, but likely one that has the replication permissions (you can find these on the replication pages).

For External IDP - this would have to be an external user with the assigned permissions.

There might need to be a Requirements section that specifically states what is needed here similar to the [bucket replication](https://docs.min.io/minio/baremetal/replication/enable-server-side-two-way-bucket-replication.html#required-permissions) pages.

Reviewer: Actually for external IDP wouldn't this need to be root? There's no "access key" or "secret key" for an AD/LDAP or OIDC user outside of either STS (which expires) or a service account (Which we have not guided the user to create).

Using root is a little bit of a security thing but that's on users to secure with a proper password, and IIRC root can differ after initial sync. Might be worth confirming w/ engineer what the best practice here is.

         :Secret Key: `(required)` The password for the user name to use for signing in to each site. Should be the same across all sites.

>          :Site Name: An alias to use for the site name.

Reviewer: Since we have some crossover of "alias" as something specific to `mc`, "A name or other identifying text to associate to the site" would be fine.

         :Endpoint: `(required)` The URL or IP address and port to use to access the site.

>          To add additional sites beyond two, select the ``+`` button to the side of one of the Site entries.

Reviewer: beyond three?

Author: Nope, two is the minimum required for site replication. I've corrected the other places where I suggested you needed three.

      
         For example, for three MinIO sites, you might create aliases ``minio1``, ``minio2``, and ``minio3``.
         
         Use :mc-cmd:`mc alias set`
      
         .. code-block:: shell
      
            mc alias set minio1 https://minio1.example.com:9000 adminuser adminpassword
            mc alias set minio2 https://minio2.example.com:9000 adminuser adminpassword
            mc alias set minio3 https://minio3.example.com:9000 adminuser adminpassword
      
         or define environment variables
      
         .. code-block:: shell
         
            export MC_HOST_minio1=https://adminuser:adminpassword@minio1.example.com
            export MC_HOST_minio2=https://adminuser:adminpassword@minio2.example.com
            export MC_HOST_minio3=https://adminuser:adminpassword@minio3.example.com
      
      #. Add site replication configuration
      
         .. code-block:: shell
         
            mc admin replicate add minio1 minio2 minio3
      
      #. Query the site replication configuration to verify
      
         .. code-block:: shell
         
            mc admin repicate info minio1

Expand site replication
-----------------------

> If necessary or desired, you can add more sites to an existing site replication.

Reviewer: ```suggestion
You can add more sites to an existing site replication configuration.
```

> The site to add must already be deployed, and it must be empty (no buckets or objects).

Reviewer: We should avoid passives whenever possible.

```suggestion
The new site must meet the following requirements:

- Site is fully deployed and accessible via hostname or IP
- Site shares the IDP configuration as all other sites in the replica configuration
- Site contains no bucket or object data
```

      
         or define environment variables
      
         .. code-block:: shell
         
            export MC_HOST_minio1=https://adminuser:adminpassword@minio1.example.com
            export MC_HOST_minio2=https://adminuser:adminpassword@minio2.example.com
            export MC_HOST_minio3=https://adminuser:adminpassword@minio3.example.com
      
      #. Add site replication configuration

         List all existing replicated sites first, then list the new site(s) to add.
         In this example, ``minio1``, ``minio2``, and ``minio3`` are already configured for replication.
         The command adds minio4 and minio5 as new sites to add to the replication.
         ``minio4`` and ``minio5`` must be empty.
      
         .. code-block:: shell
         
            mc admin replicate add minio1 minio2 minio3 minio4 minio5
      
      #. Query the site replication configuration to verify
      
         .. code-block:: shell
         
            mc admin replicate info minio1

Modify a site's endpoint
------------------------

From time to time a replicated site's address may change.

> When this occurs, update the site's endpoint in the site replication's configuration.

Reviewer: Personal nit - we don't need the soft lead in, we can get straight to the point.

"If a site in the replication config has changed or will change its hostname, you can modify the replication configuration to reflect the new hostname."

            :width: 600px
            :alt: MinIO Console's List of Replicated Sites screen with the edit buttons highlighted
            :align: center

      #. Make the following entries:

         :New Endpoint: `(required)` The new endpoint address and port to use.

         .. image:: /images/minio-console/console-settings-site-replication-edit-endpoint.png
            :width: 600px
            :alt: Example of the MinIO Console's Edit Replication Endpoint screen
            :align: center

      #. Click **Update**

   .. tab-item:: Command line

      #. Obtain the site's Deployment ID with :mc-cmd:`mc admin replicate info`

         .. code-block:: shell

            mc admin replicate info <ALIAS>
         
      
      #. Update the site's endpoint with :mc-cmd:`mc admin replicate edit`
      
         .. code-block:: shell

            mc admin replicate edit ALIAS --deployment-id [DEPLOYMENT-ID] --endpoint [NEW-ENDPOINT]

         Replace [DEPLOYMENT-ID] with the deployment ID of the site to update.
         
         Replace [NEW-ENDPOINT] with the new endpoint for the site.

Remove a Site from replication
------------------------------

> When decomissioning a peer site, you must remove it from the site replication configuration.

Reviewer: Decommissioning here could be misconstrued with pool decommissioning. Maybe just "You can remove a site from the replication configuration at any time. You cannot re-add the site at a later date without first completely wiping data from that site and re-adding it."

Assuming the second part based on the whole deployment ID thing. Engineering can confirm.

      #. Select **Settings**, then **Site Replication**
      
      #. Select the trash can Delete icon to the side of the site to update

         .. image:: /images/minio-console/console-site-replication-delete-button.png
            :width: 600px
            :alt: MinIO Console's List of Replicated Sites screen with the delete buttons highlighted
            :align: center

      #. Confirm the site deletion at the prompt by selecting **Delete**

         .. image:: /images/minio-console/console-settings-site-replication-confirm-delete.png
            :width: 600px
            :alt: Example of the MinIO Console's Edit Replication Endpoint screen
            :align: center

   .. tab-item:: Command line

      Use :mc-cmd:`mc admin replicate remove`

      .. code-block:: shell

         mc admin replicate remove <ALIAS> --all --force

      The ``-all`` flag removes the site as a peer from all participating sites.

      The ``--force`` flag is required to removes the site from the site replication configuration.

Review replication status
-------------------------

MinIO provides information on replication across the sites.

> Information is available in a summary view or for specific, inlcuding users, groups, policies, or buckets.

Reviewer: This could be one concise paragraph.

      
      #. Select :guilabel:`Replication Status`

         .. image:: /images/minio-console/console-settings-site-replication-status-summary.png
            :width: 600px
            :alt: MinIO Console's Replication status from all Sites screen
            :align: center

      #. `(Optional)` View the replication status for a specific item
         
         Select the type of item to view in the :guilabel:`View Replication Status for a:` dropdown

         Specify the name of the specific Bucket, Group, Policy, or User to view

         .. image:: /images/minio-console/console-settings-site-replication-status-item.png
            :width: 600px
            :alt: Example of replication status for a particular bucket item
            :align: center
      
      #. `(Optional)` Update the information by clicking :guilabel:`Refresh`

   .. tab-item:: Command line

      Use :mc-cmd:`mc admin replicate status`

      .. code-block:: shell

         mc admin replicate status <ALIAS> --<flag> <value>

      For example:

      - ``mc admin replicate status minio3 --bucket images``

        Displays the replication status for the ``images`` bucket on the ``minio3`` site.
        

>         Example output:

Reviewer: The output resembles the following <- buys us CYA in case the output format changes slightly.

Also as obvious as it should be, we should note what checkmark, x-mark (I assume), and a  blank entry mean

.. _minio-site-replication-overview:

=========================
Site Replication Overview
=========================

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 1

Site replication configures multiple independent MinIO deployments as a cluster of replicas called peer sites.

> Site replication assumes the use of an the included MinIO identity provider (IDP) or an external IDP.

Author: done

> All configured deployments must use the same IDP.

Author: Updated

.. _minio-site-replication-what-replicates:

> What Replicates Across All Sites

Author: This is a good suggestion. Applied.

> What Replicates Across All Sites

Author: If we want to pull out the tutorials later, we can ask a separate issue.

--------------------------------

Each MinIO deployment ("peer site") synchronizes the following changes across the other peer sites:

- Creation, modification, and deletion of buckets and objects, including

  - Bucket and Object Configurations
  - :ref:`Policies <minio-policy>`
  - :mc-cmd:`mc tag`
  - :ref:`Locks <minio-object-locking>`, including retention and legal hold configurations
  - :ref:`Encryption settings <minio-encryption-overview>`

- Creation and deletion of IAM users, groups, policies, and policy mappings to users or groups (for LDAP users or groups)
- Creation of Security Token Service (STS) credentials for session tokens verifiable from the local ``root`` credentials
- Creation and deletion of :ref:`service accounts <minio-mc-admin-user-svcacct>` (except those owned by the ``root`` user)

Site replication enables :ref:`bucket versioning <minio-bucket-versioning>` for all new and existing buckets on all replicated sites.

What Does Not Replicate Across Sites
------------------------------------

> The following bucket features don't replicate across sites set up for site replication:

Author: Done

- :ref:`Bucket notifications <minio-bucket-notifications>`
- :ref:`Lifecycle management (ILM) configurations <minio-lifecycle-management>`
- :ref:`Site configuration settings <minio-mc-admin-config>`

> This allows users who access different sites to define separate site settings, bucket configurations, and bucket notifications after the initial synchronization completes.

Author: Deleted.

Order of Operations
-------------------

After enabling site replication, identity and access management (IAM) settings sync in the following order:

.. tab-set::

   .. tab-item:: MinIO IDP

      #. Policies
      #. User accounts (for local users)
      #. Groups
      #. Service accounts
         

>          Only service accounts related to user accounts with valid claims.

Author: Removed from the MinIO IDP tab.
Updated to the following for the OIDC and LDAP tabs:
```
Only service accounts related to user accounts with valid policies for the MinIO deployment.
```

      #. Policy mapping for `Security Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

   .. tab-item:: OIDC

      #. Policies
      #. Service accounts
         
         Only service accounts related to user accounts with valid claims.
         Service accounts for ``root`` do not sync.

      #. Policy mapping for synced user accounts
      #. Policy mapping for `Security Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

   .. tab-item:: LDAP

      #. Policies
      #. Groups
      #. Service accounts

         Only service accounts related to user accounts with valid claims.
         Service accounts for ``root`` do not sync.

      #. Policy mapping for synced user accounts
      #. Policy mapping for `Security Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

Site Healing
------------

With site replication enabled, for any item :ref:`supported in site replication <minio-site-replication-what-replicates>`, the MinIO deployment resynchronizes that item from the peer site with the most updated ("latest") version of that item.

Prerequisites
-------------

> One Site with Data at setup

Author: Le sigh. Fixed.

~~~~~~~~~~~~~~~~~~~~~~~~~~~

Only *one* site can have data at the time of setup.
The other sites must be empty of buckets and objects.

After configuring site replication, any data  on the first deployment replicates to the other sites.
After the initial replication completes, changes to :ref:`items that replicate <minio-site-replication-what-replicates>` on any of the peer sites synchronize to all of the other peer sites.
Changes include creating, modifying, or deleting such items.

> 

Author: Added admonition

All Sites Must Use the Same Root Deployment Credentials
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

All sites must have the same root user credentials (for example, ``MINIO_ROOT_USER``, ``MINIO_ROOT_PASSWORD``).

> This allows MinIO to log in across each site.

Reviewer 2: all sites are not required to have same root credentials. When setting up site replication, service account  will be created automatically to handle site replication activity.

> This allows MinIO to log in across each site.

Author: Removed this prereq

All Sites Must Use the Same IDP
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> All sites must use the same Identity Provider.

Author: Done.

Site replication supports the included MinIO IDP, OIDC, or LDAP.

Access to the Same Encryption Service
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For :ref:`SSE-S3 <minio-encryption-sse-s3>` or :ref:`SSE-KMS <minio-encryption-sse-kms>` encryption via Key Management Service (KMS), all sites must have access to a central KMS deployment. 

> You can achieve this with a central KES server or multiple KES servers (say one per site) connected via a central KMS (Vault) server.

Author: Reworded and added a link to the MinIO SSE doc.

> Configure site replication

Author: :|

--------------------------

.. tab-set::

   .. tab-item:: Console

      #. :ref:`Deploy <deploy-minio-distributed>` two or more separate MinIO sites, using the same Identity Provider for each site

         Only one site can have any buckets or objects on it.
         The other site(s) must be empty.

      #. In a browser, access the Console for the site with data (if any)

>          .. important:: 

Author: Agreed

         
            If one of the sites has data, you must log in to the Console for that site.

         For example, ``https://<addressforsite>:9000``
         
         Replace ``<addressforsite>`` with the IP address or URL for the MinIO deployment.

      #. Select **Settings**, then **Site Replication**

         .. image:: /images/minio-console/console-settings-site-replication.png
            :width: 400px
            :alt: MinIO Console menu with the Settings heading expanded to show Site Repilication
            :align: center
      

>       #. Select :guilabel:`Add Sites +``

Author: fixed

         .. image:: /images/minio-console/console-settings-site-replication-add.png
            :width: 600px
            :alt: MinIO Console's Add Sites for Replication screen
            :align: center

      #. Complete the requested information for the site:

>          :Access Key: `(required)` The user name for ``root`` to use for signing in to each site. Must be the same across all sites.

Author: It's in the rST docs as a field list. Thought it made sense to use here. 
Supports alternate colored rows in the table it creates, too.

         :Secret Key: `(required)` The password for ``root`` to use for signing in to each site. Must be the same across all sites.

>          :Site Name: An name or other identifying text to associate to the site.

Author: Fixed

         :Endpoint: `(required)` The URL or IP address and port to use to access the site.

         To add additional sites beyond two, select the ``+`` button to the side of one of the Site entries.
         To remove a site previously added, select the ``-`` button to the side of the site.

>       #. Click **Save**

Author: Added a step to review the status for both the console and the CLI.

   .. tab-item:: Command line

      #. Deploy three or more separate MinIO sites, using the same external IDP

>          Only one site can have any buckets or objects on it.

Author: Moved the order of aliases to the step for `mc admin replicate add`. Clarified that all sites can be empty, or no more than one site can contain data.

         The other sites must be empty.

      #. Configure an alias for each site
      

>          For example, for three MinIO sites, you might create aliases ``minio1``, ``minio2``, and ``minio3``.

Author: Set up the scenario prior to the first step.

         
         Use :mc-cmd:`mc alias set`
      
         .. code-block:: shell
      
            mc alias set minio1 https://minio1.example.com:9000 adminuser adminpassword
            mc alias set minio2 https://minio2.example.com:9000 adminuser adminpassword
            mc alias set minio3 https://minio3.example.com:9000 adminuser adminpassword
      
         or define environment variables
      
         .. code-block:: shell
         
            export MC_HOST_minio1=https://adminuser:adminpassword@minio1.example.com
            export MC_HOST_minio2=https://adminuser:adminpassword@minio2.example.com
            export MC_HOST_minio3=https://adminuser:adminpassword@minio3.example.com
      
      #. Add site replication configuration
      
         .. code-block:: shell
         
            mc admin replicate add minio1 minio2 minio3
      
      #. Query the site replication configuration to verify
      
         .. code-block:: shell
         

>             mc admin repicate info minio1

Author: Done

         For example, ``https://<addressforsite>:9000``

      #. Select **Settings**, then **Site Replication**

         .. image:: /images/minio-console/console-site-replication-list-of-sites.png
            :width: 600px
            :alt: MinIO Console Site Replication with three sites listed
            :align: center
      
      #. Select :guilabel:`Add Sites +`

         .. image:: /images/minio-console/console-settings-site-replication-add.png
            :width: 600px
            :alt: MinIO Console's Add Sites for Replication screen
            :align: center

      #. Make the following entries:

         :Access Key: `(required)` The user name to use for signing in to each site. Should be the same across all sites.

         :Secret Key: `(required)` The password for the user name to use for signing in to each site. Should be the same across all sites.

         :Site Name: An alias to use for the site name.

         :Endpoint: `(required)` The URL or IP address and port to use to access the site.

         To add additional sites beyond two, select the ``+`` button to the side of the last Site entry.

      #. Click :guilabel:`Save`

>    .. tab-item:: Command line

Author: Updated throughout

All configured deployments must use the same IDP.
Deployments using an external IDP must use the same configuration across sites.

Overview
--------

.. _minio-site-replication-what-replicates:

What Replicates Across All Sites
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. include:: /includes/common-replication.rst
   :start-after: start-mc-admin-replicate-what-replicates
   :end-before: end-mc-admin-replicate-what-replicates

What Does Not Replicate Across Sites
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. include:: /includes/common-replication.rst
   :start-after: start-mc-admin-replicate-what-does-not-replicate
   :end-before: end-mc-admin-replicate-what-does-not-replicate

Initial Site Replication Process
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

After enabling site replication, identity and access management (IAM) settings sync in the following order:

.. tab-set::

   .. tab-item:: MinIO IDP

      #. Policies
      #. User accounts (for local users)
      #. Groups
      #. Service accounts
         
         Service accounts for ``root`` do not sync.

      #. Policy mapping for synced user accounts

>       #. Policy mapping for `Security Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

Author: NICE catch

Overview
--------

.. _minio-site-replication-what-replicates:

What Replicates Across All Sites
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. include:: /includes/common-replication.rst
   :start-after: start-mc-admin-replicate-what-replicates
   :end-before: end-mc-admin-replicate-what-replicates

What Does Not Replicate Across Sites
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. include:: /includes/common-replication.rst
   :start-after: start-mc-admin-replicate-what-does-not-replicate
   :end-before: end-mc-admin-replicate-what-does-not-replicate

Initial Site Replication Process
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

After enabling site replication, identity and access management (IAM) settings sync in the following order:

.. tab-set::

   .. tab-item:: MinIO IDP

      #. Policies
      #. User accounts (for local users)
      #. Groups
      #. Service accounts
         
         Service accounts for ``root`` do not sync.

      #. Policy mapping for synced user accounts
      #. Policy mapping for `Security Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

   .. tab-item:: OIDC

      #. Policies

>       #. Service accounts
>          
>          Only service accounts related to user accounts with valid policies for the MinIO deployment.
>          Service accounts for ``root`` do not sync.
> 

Reviewer: This renders a little weird with the paragraph break - I wonder if we can simplify this:

```suggestion
      #. 
         Service accounts associated to OIDC accounts with valid :ref:`MinIO policy <minio-policy>`. ``root`` service accounts do not sync.

```

Or something

Author: Sure. I just used your suggestion for both OIDC and LDAP.

      #. Policy mapping for synced user accounts
      #. Policy mapping for `Security Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

   .. tab-item:: LDAP

      #. Policies
      #. Groups
      #. Service accounts

         Only service accounts related to user accounts with valid policies for the MinIO deployment.
         Service accounts for ``root`` do not sync.

      #. Policy mapping for synced user accounts
      #. Policy mapping for `Security Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

> 

Reviewer: I think we should add something like:

```suggestion
After initial synchronization of data completes, MinIO continually replicates and synchronizes :ref:`replication-eligible changes <>` as they occur.
```

With a label to the section on "What replicates"

Author: Slight tweak, but yes.

Site Healing
~~~~~~~~~~~~

> With site replication enabled, for any item :ref:`supported in site replication <minio-site-replication-what-replicates>`, the MinIO deployment resynchronizes that item from the peer site with the most updated ("latest") version of that item.

Reviewer: I think we can slightly tweak this to simplify the structure - since we can (I think) safely assume the reader understands this is in context of site replication

```suggestion
Any MinIO deployment in the replication configuration can resynchronize damaged :ref:`replica-eligible data <minio-site-replication-what-replicates>` from the peer with the most updated ("latest") version of that data.
```

Author: I did `Any MinIO deployment in the site replication configuration` to make sure we distinguish site and bucket replication. But otherwise used your suggestion.

Prerequisites
-------------

One Site with Data at Setup
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Only *one* site can have data at the time of setup.

> The other sites must be empty of buckets and objects.

Reviewer: Would it require being empty of all replica-eligible data, or *just* bucket/objects? e.g. if two sites have IAM data persisted, what happens.

Author: Confirmed that it is just buckets and objects.
IAM data can exist and differ between the sites prior to enabling site replication, and replication occurs just fine (including sharing all IAM data between sites, for example).

After configuring site replication, any data on the first deployment replicates to the other sites.

After the initial replication completes, changes to :ref:`items that replicate <minio-site-replication-what-replicates>` on any of the peer sites synchronize to all of the other peer sites.

> Changes include creating, modifying, or deleting such items.

Reviewer: This feels out of place given that this is a prereq section and we're discusssing behavior here.

I think this is where you intended to have the admonition warning against reading/writing from/to peer sites until replication completes?

Author: Yes, and Poorna said that was unnecessary.
I deleted these two lines.

      #. Policy mapping for `Security Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

   .. tab-item:: LDAP

      #. Policies
      #. Groups
      #. Service accounts

         Only service accounts related to user accounts with valid policies for the MinIO deployment.
         Service accounts for ``root`` do not sync.

      #. Policy mapping for synced user accounts
      #. Policy mapping for `Security Token Service (STS) users <https://docs.min.io/docs/minio-sts-quickstart-guide.html>`

Site Healing
~~~~~~~~~~~~

With site replication enabled, for any item :ref:`supported in site replication <minio-site-replication-what-replicates>`, the MinIO deployment resynchronizes that item from the peer site with the most updated ("latest") version of that item.

Prerequisites
-------------

One Site with Data at Setup
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Only *one* site can have data at the time of setup.
The other sites must be empty of buckets and objects.

After configuring site replication, any data  on the first deployment replicates to the other sites.

.. admonition:: Wait Until Replication Completes
   :class: warning

>    If there is existing data on the first peer site, wait until :mc-cmd:`mc admin replicate status` shows that the replication process has completed before performing read or write operations on the other peer sites. 

Reviewer 2: not necessary to wait - if other sites are missing data, proxying will take care of reading from the first site

Author: Removed the admonition.

         Only one site can have any buckets or objects on it.
         The other site(s) must be empty.

      #. In a browser, access the Console for the site with data (if any)

         For example, ``https://<addressforsite>:9000``
         
         Replace ``<addressforsite>`` with the IP address or URL for the MinIO deployment.

      #. Select **Settings**, then **Site Replication**

         .. image:: /images/minio-console/console-settings-site-replication.png
            :width: 400px
            :alt: MinIO Console menu with the Settings heading expanded to show Site Repilication
            :align: center
      
      #. Select :guilabel:`Add Sites +`

         .. image:: /images/minio-console/console-settings-site-replication-add.png
            :width: 600px
            :alt: MinIO Console's Add Sites for Replication screen
            :align: center

      #. Complete the requested information for the site:

         :Access Key: `(required)` The user name for ``root`` to use for signing in to each site. Must be the same across all sites.

>          :Secret Key: `(required)` The password for ``root`` to use for signing in to each site. Must be the same across all sites.

Reviewer 2: root access key and secret key need not be identical across sites.

Author: Removed.
